pkg/storage/etcd3/etcdwatcher/watcher.go

```diff
@@ -0,0 +1,340 @@
+	// Watch watches on a key and returns a watch.Interface that transfer relevant notifications.
+	// If rev is zero, it will try to get existing objects if any and then watch after that revision.
+	// If rev is non-zero, it will watch from given revision.
+	// If recursive is false, it watches on given key.
+	// If recursive is true, it watches any children and directories under the key, excluding the root key itself.
+	// filter must be non-nil. Only if filter returns true will the changes be returned.
+	Watch(ctx context.Context, key string, rev int64, recursive bool, filter storage.FilterFunc) (watch.Interface, error)
+}
+
+// New returns a Watcher.
+func New(client *clientv3.Client, codec runtime.Codec, versioner storage.Versioner) Watcher {
+	return newWatcher(client, codec, versioner)
+}
+
+// watcher implements Watcher.
+type watcher struct {
+	client    *clientv3.Client
+	codec     runtime.Codec
+	versioner storage.Versioner
+}
+
+// watchChan implements watch.Interface.
+type watchChan struct {
+	watcher       *watcher
+	key           string
+	initialRev    int64
+	recursive     bool
+	filter        storage.FilterFunc
+	ctx           context.Context
+	cancel        context.CancelFunc
+	resultChan    chan watch.Event
+	eventChan     chan *event
+	resultQueue   *list.List
+	errChan       chan error
+	waitSync      chan struct{}
+	waitEventProc sync.WaitGroup
+}
+
+func newWatcher(client *clientv3.Client, codec runtime.Codec, versioner storage.Versioner) *watcher {
+	return &watcher{
+		client:    client,
+		codec:     codec,
+		versioner: versioner,
+	}
+}
+
+func (w *watcher) Watch(ctx context.Context, key string, rev int64, recursive bool, filter storage.FilterFunc) (watch.Interface, error) {
+	if recursive && !strings.HasSuffix(key, "/") {
+		key += "/"
+	}
+	wc := &watchChan{
+		watcher:     w,
+		key:         key,
+		initialRev:  rev,
+		recursive:   recursive,
+		filter:      filter,
+		resultChan:  make(chan watch.Event, 1),
+		eventChan:   make(chan *event, 100),
+		resultQueue: list.New(),
+		errChan:     make(chan error, 1),
+		waitSync:    make(chan struct{}),
+	}
+	wc.ctx, wc.cancel = context.WithCancel(ctx)
+	go wc.run()
+	wc.waitEventProc.Add(1)
+	return wc, nil
+}
+
+func (wc *watchChan) run() {
+	go wc.sync()
+	go wc.startWatching()
+	go wc.eventProcessor()
+
+	var errResult *watch.Event
+	select {
+	case err := <-wc.errChan:
+		errResult = wc.parseError(err)
+		wc.cancel()
+	case <-wc.ctx.Done():
+	}
+	// error result is guaranteed to be received by user before closing ResultChan.
+	if errResult != nil {
+		wc.resultChan <- *errResult
+	}
+	wc.waitEventProc.Wait()
+	close(wc.resultChan)
+}
+
+func (wc *watchChan) Stop() {
+	wc.cancel()
+}
+
+func (wc *watchChan) ResultChan() <-chan watch.Event {
+	return wc.resultChan
+}
+
+// sync tries to retrieve existing data send them to process.
+// The revision to watch will be set to the revision in response.
+func (wc *watchChan) sync() {
+	defer close(wc.waitSync)
```

Comment:
This pattern is quite unnatural.
What we should do is to have a single function that is run in go-routine (e.g. watchFromEtcd or syncFromEtcd or sth like that) and in this function you should have:

if wc.initialRev == 0 {
  // get & send all initial elements
  wc.sendInitialEvents()
}

The threading pattern here is very difficult to follow and error-prone.